### PR TITLE
Hackney adapter

### PR DESCRIPTION
**Pull request by keathley**
This PR adds a basic hackney adapter.

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -46,6 +46,7 @@ defmodule Twirp.MixProject do
       {:protobuf, "~> 0.5"},
       {:google_protos, "~>0.1"},
       {:finch, "~> 0.3"},
+      {:hackney, "~> 1.17"},
```
Comment (wojtekmach) [inline mix.exs:49]: consider making this dependency `optional: true` so that people that don't use it won't get it (well, they'd likely get it anyway from some other dep :))

```elixir
defmodule Twirp.Client.Hackney do
  @impl true
  def init() do
    if Code.ensure_loaded?(:hackney) do
      Logger.error("""
      Could not find hackney dependency.

      Please add :hackney to your dependencies:

          {:hackney, "~> 1.17"}
      """)

      raise "missing hackney dependency"
    else
      {:ok, _} = Application.ensure_all_started(:hackney)
      :ok
    end
  end

  @impl true
  def request(...)
end
```

I'm happy to prepare a PR after this lands or to this branch if you prefer!

Comment (keathley) [reply mix.exs]: I just pushed a quick change for this. Can you double check me that I didn't introduce something weird? I think its correct but some of this `Code.ensure_loaded` stuff trips me up occasionally.

File: lib/twirp/client/finch.ex [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    if Code.ensure_loaded?(Finch) do
+      opts = Keyword.new(opts)
+      Finch.start_link(opts)
+    else
+      raise AdapterError, :finch
+    end
+  end
+
+  def request(client, ctx, path, payload) do
+    if Code.ensure_loaded?(Finch) do
```
Comment (wojtekmach) [inline lib/twirp/client/finch.ex:17]: I'm told that `Code.ensure_loaded?` should be super fast on recent OTPs but I'd still avoid using that in "hot path" which I'd consider request/4 to be in. If we're guaranteed that start_link/1 of each adapter would be called in before we start making requests, I think doing the check just there is enough, and since it will be executed very seldom it's totally ok to use Code.ensure_loaded? there.

TLDR keep start_link/1 as is, remove this line from request/4 and I think you're good.

Comment (keathley) [reply lib/twirp/client/finch.ex:17]: I think that ends up creating a bunch of compiler warnings because the compiler detects that there are functions in there that aren't actually defined anywhere. I don't think it _breaks_ but I'm wondering if there is a clean way to handle it. I think I can make it work with `Code.ensure_compiled`.

Comment (wojtekmach) [reply lib/twirp/client/finch.ex:17]: that is weird, I've used this exact pattern a bunch. Let me play with this and get back to you.

Comment (wojtekmach) [reply lib/twirp/client/finch.ex:17]: oh, I guess it makes sense. I think the compiler is smart enough to not warn on code guarded with Code.ensure_loaded? but what I'm proposing is to leave that code unguarded.

We can suppress the warning by doing this: https://github.com/swoosh/swoosh/pull/468/files#diff-dfa6f4ed74c90e5d4fda283d547d366586e690387289bcfd473e3fa5f9ace308R35 :)

Comment (keathley) [reply lib/twirp/client/finch.ex:17]: The last time I tried guarding the functions definitions with `Code.ensure_loaded` in Norm, there was also an ordering issue where `Code.ensure_loaded?` can fail if one dependency happened to compiler before the other.

I didn't realize you could disable xref warnings like that. That seems reasonable as long as it doesn't halt the compiler or anything like that.

Comment (wojtekmach) [reply lib/twirp/client/finch.ex]: oh, that's what you meant about Code.ensure_loaded tripping you up, yeah, no idea about ensure_compiled. And yeah, disabling warnings like that is solid, we also use it e.g. in ecto_sql.
